src/litdata/streaming/downloader.py

```diff
@@ -203,6 +209,56 @@ def download_bytes(self, remote_filepath: str, offset: int, length: int, local_c
 
         return response["Body"].read()
 
+    def download_fileobj(self, remote_filepath: str, fileobj: Any) -> None:
+        """Download a file from S3 directly to a file-like object."""
+        obj = parse.urlparse(remote_filepath)
+
+        if obj.scheme != "s3":
+            raise ValueError(f"Expected obj.scheme to be `s3`, instead, got {obj.scheme} for remote={remote_filepath}")
+
+        if not hasattr(self, "_client"):
+            self._client = S3Client(storage_options=self._storage_options, session_options=self.session_options)
+
+        bucket = obj.netloc
+        key = obj.path.lstrip("/")
+
+        self._client.client.download_fileobj(
+            bucket,
+            key,
+            fileobj,
+        )
+
+    def _get_store(self, bucket: str):
+        """Return an obstore S3Store instance for the given bucket, initializing if needed."""
+        if not hasattr(self, "_store"):
+            if not _OBSTORE_AVAILABLE:
+                raise ModuleNotFoundError(str(_OBSTORE_AVAILABLE))
+            from obstore.auth.boto3 import Boto3CredentialProvider
+            from obstore.store import S3Store
+
+            session = boto3.Session(**self._storage_options, **self.session_options)
+            credential_provider = Boto3CredentialProvider(session)
+            self._store = S3Store(bucket, credential_provider=credential_provider)
+        return self._store
+
+    async def adownload_fileobj(self, remote_filepath: str, fileobj: Any) -> None:
+        """Download a file from S3 directly to a file-like object asynchronously."""
+        import obstore as obs
+
+        obj = parse.urlparse(remote_filepath)
+
+        if obj.scheme != "s3":
+            raise ValueError(f"Expected obj.scheme to be `s3`, instead, got {obj.scheme} for remote={remote_filepath}")
+
+        bucket = obj.netloc
+        key = obj.path.lstrip("/")
+
+        store = self._get_store(bucket)
+        resp = await obs.get_async(store, key)
+        stream = resp.stream(min_chunk_size=2 * 1024 * 1024)
```

Comment:
I think we should use https://developmentseed.org/obstore/latest/api/get/#obstore.GetResult.buffer_async instead. So we let rust get the bytes ready and then there is a no copy back to python

Reply:
<img width="1466" height="202" alt="Screenshot 2025-07-25 at 10 38 09" src="https://github.com/user-attachments/assets/d2dda47b-d346-4018-90f4-ec454d7423fb" />